core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -208,6 +209,10 @@ class TransactionCoordinator(brokerId: Int,
     if (transactionalId == null || transactionalId.isEmpty) {
       responseCallback(Errors.INVALID_REQUEST)
     } else {
+      // if there is any partitions unknown in the metadata cache, return immediately to client
+      if (partitions.exists(tp => !metadataCache.contains(tp.topic, tp.partition)))
+        responseCallback(Errors.UNKNOWN_TOPIC_OR_PARTITION)
```

Comment:
Thinking about this a bit more, the behavior of the client will be tricky. Essentially, the send will block until it's expired, and yet the `AddPartitions` will keep retrying indefinitely. And because of https://issues.apache.org/jira/browse/KAFKA-5251, the client will keep retrying the `AddPartitions` even if the user aborts. If we want to keep this behavior, I think we may want to fix KAFKA-5251 as well. This way the user can at least abort the transaction properly after realizing that they are trying to send to deleted partitions.